Actions cheatsheet

brntbeer opened the pull request:
## Overview
**TL;DR**
This is a new cheat sheet for [GitHub Actions](https://github.com/features/action) as mentioned in https://github.com/github/training-kit/issues/691

### Review
not ready yet but cc @helaili @johndavidmarx @hectorsector and @mattdavis0351 anddd @brianamarie  for general heads up

brntbeer commented on downloads/github-actions-cheat-sheet.md:
```diff
@@ -0,0 +1,209 @@
+### `jobs`
+A workflow run is made up of one or more jobs. Jobs run in parallel by default. To run jobs sequentially, you can define dependencies on other jobs using the `jobs.<job_id>.needs` keyword. Each job runs in a fresh instance of the virtual environment specified by runs-on.
+
+### `runs-on`
```
One thing we may want to do with commands after our jobs heading is make them level 4 headers #### because it continues the hierarchy that is set up in the yaml. runs-on for instance is actually jobs.<job_id>.runs-on in the documentation.

Thoughts?

brntbeer replied:
it ends up looking nicer with `<h4>` headings. marking this as resolved.

brntbeer commented on downloads/github-actions-cheat-sheet.md:
```diff
@@ -0,0 +1,209 @@
+
+- `ubuntu-latest`, `ubuntu-18.04`, or `ubuntu-16.04`
+- `windows-latest`, `windows-2019`, or `windows-2016`
+- `macOS-latest` or `macOS-10.14`
```
I'm a bit apprehensive for adding versions here because this makes futureproofing it real difficult. I'd like to show something like just the latest versions.

bryancross replied:
Agreed, include a link to the relevant docs section for latest info

brntbeer commented:
![Screen Shot 2019-09-10 at 11 40 01 AM](https://user-images.githubusercontent.com/45141/64602849-c347fd00-d3bf-11e9-9886-297fdf99a57f.png)

(_Sheepishly_) Page 1 ready for review

brntbeer commented:
The more i look at the first page, I'm wondering if we need to include a full yaml example that's as minimum as possible for people to refer to. This would likely push everything from `jobs` downward into a page 2 column 1. whichhhh may work out if we focus on page 2 column 2 to just be `strategies` `secrets` and `variable` usage?

hectorsector commented:
> I'm wondering if we need to include a full yaml example that's as minimum as possible for people to refer to

I'm 👍 for this. I always struggle with documentation that makes the structure appear flat. A sample YAML, or anything that shows you the parent/child/sibling relationships of the syntax would be hugely beneficial.

brianamarie commented:
This looks awesome so far, @brntbeer! (at least from my perspective of someone who hasn't written actions successfully yet.)  ✨ 


One thing that has been confusing to me as I try to get actions working is the differentiation between a _workflow_ and an _action_. I was trying for a long time to write the action in the same file or directory as the workflow, rather than create an action in a separate part of the repository and then reference it in the workflow file. What are your thoughts on adding something to define workflows vs actions, or some visual to help communicate the roles played by those two components? 

> I'm wondering if we need to include a full yaml example that's as minimum as possible for people to refer to

+1 to full example of yaml file

brntbeer commented:
- Updated to include the section header "workflow syntax" instead of just "syntax" (cc @brianamarie)
- unfortunately, the workflow syntax spills into the second page now (see deployment: https://github-f033955105.drafts.github.io/training-kit/downloads/github-actions-cheat-sheet/)
  - @helaili @brianamarie and anyone else: thoughts on this? My feeling is kind of to create a `##` header on the "third column" (first column second page) and say "workflow syntax cont."
- removed individual syntax examples for uses/name and runs.

Ideally, what's left now is to:
- figure out syntax highlighting in this repo for yaml (or just all code examples)
- describe `strategies`, `secrets`, and pattern matching as a more advance feature

brianamarie commented:
> Updated to include the section header "workflow syntax" instead of just "syntax" 

🙏 Thank you!

> * My feeling is kind of to create a `##` header on the "third column" (first column second page) and say "workflow syntax cont."

👍 I think this is a good idea.

helaili commented:
> unfortunately, the workflow syntax spills into the second page now

How can you tell @brntbeer ?

brntbeer commented:
> How can you tell @brntbeer ?

not an exact science but based on page lengths from prior columns and keeping the length the same

helaili commented:
[Current's version PDF rendering](https://github.com/github/training-kit/files/3661438/github-action-cheat-sheet-V1.0.pdf)

kai commented:
Requested reviews from the `@github/smg-actions` team

bryancross commented on downloads/github-actions-cheat-sheet.md:
````diff
@@ -0,0 +1,193 @@
+```
+
+### Workflow `name`
+The name of your workflow. GitHub displays the names of your workflows on your repository's actions page.
````
Are there any naming rules? Forbidden characters?  Nothing in the docs but I wonder if special chars are forbidden.

bryancross commented on downloads/github-actions-cheat-sheet.md:
````diff
@@ -0,0 +1,193 @@
+```
+
+### Workflow `name`
+The name of your workflow. GitHub displays the names of your workflows on your repository's actions page.
````
```suggestion
The name of your workflow. GitHub displays the names of your workflows on your repository's actions page.  Defaults to your workflow filename if omitted.
```

helaili replied:
We are aiming for a 2 pager so we need to balance completness with real estate optmization.

bryancross commented on downloads/github-actions-cheat-sheet.md:
````diff
@@ -0,0 +1,193 @@
+- The `types` keyword enables you to narrow down activity that causes the workflow to run: `created`, `edited`, `deleted`...
+A workflow trigger can also be scheduled:  
+```yaml
+on:
````
The schedule example here seems like a non-sequitur since it isn't mentioned in the text above.  Perhaps include an example mentioned therein.  Also, I think it's worth highlighting that you can filter by `path` as well as `paths-ignore` (although not in the same job) . 

If we mention the `schedule` option, consider adding a link to the [syntax explanation](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onschedule)

helaili replied:
Absolutely. Something disappeard here and new options have been released since it was written. The whole section need a refresh.

bryancross commented on downloads/github-actions-cheat-sheet.md:
```diff
@@ -0,0 +1,193 @@
+Identifies any job that must complete successfully before this job will run. It can be a string or array of strings. If a job fails, all jobs that need it are skipped unless the jobs use a conditional statement that causes the job to continue.
+
+#### `runs-on`
+Inside of the `jobs` collection, this is the type of virtual host machine to run the job on.
```
```suggestion
The type of virtual host machine to run the job on.  Can be either a GitHub or self-hosted runner.  Jobs can also run in user-specified containers (see: `container`)
```

bitoiu replied:
I was going to edit the `inside of the job` but I see @bryancross already proposed a change so :+1:

bryancross commented on downloads/github-actions-cheat-sheet.md:
```diff
@@ -0,0 +1,193 @@
+Specify the label to be displayed for this step in GitHub. It's not required but does improve readability in the logs.
+
+#### `uses`
+Select an action to run as part of a step in your job. You can use an action defined in the same repository as the workflow, a public repository elsewhere on GitHub, or in a published Docker container image.
```
```suggestion
Specify an action to run as part of a step in your job. You can use an action defined in the same repository as the workflow, a public repository elsewhere on GitHub, or in a published Docker container image.
```

bryancross commented on downloads/github-actions-cheat-sheet.md:
```diff
@@ -0,0 +1,193 @@
+Select an action to run as part of a step in your job. You can use an action defined in the same repository as the workflow, a public repository elsewhere on GitHub, or in a published Docker container image.
+
+#### `with`
+A map of the input parameters defined by the action. Each input parameter is a key/value pair. Input parameters are prefixed with `INPUT_`, converted to upper case and set as environment variables: the `first_name` parameter will be exposed as the `INPUT_FIRST_NAME` environment variable. Special parameter names are:
```
```suggestion
A map of the input parameters defined by the action in its `action.yml` file. Each input parameter is a key/value pair. Input parameters are prefixed with `INPUT_`, converted to upper case and set as environment variables: the `first_name` parameter will be exposed as the `INPUT_FIRST_NAME` environment variable. Special parameter names are:
```

bryancross commented on downloads/github-actions-cheat-sheet.md:
```diff
@@ -0,0 +1,193 @@
+- `entrypoint`, a string that defines or overrides the executable to run as the Docker container's `ENTRYPOINT`.
+
+#### `env`
+Sets environment variables for steps to use in the virtual environment.  If you are setting a secret in an environment variable, you must use the `secrets` context.
```
A word on whether this is a safe practice would be helpful here.

helaili replied:
Not sure what to say here and whether it fits on a cheat sheet

bryancross commented on downloads/github-actions-cheat-sheet.md:
````diff
@@ -0,0 +1,193 @@
+      last_name: ${{secrets.LASTNAME}}
+```
+
+Default environment variables:
````
Missing `GITHUB_ACTIONS`

helaili replied:
I'm thinking about shrinking this section and keeping essential variables only

bryancross commented on downloads/github-actions-cheat-sheet.md:
```diff
@@ -0,0 +1,193 @@
+`GITHUB_HEAD_REF`: Branch of the head repository (forks only).
+`GITHUB_BASE_REF`: Branch of the base repository (forks only).
+
+#### `run`
```
can run multiple commands in a single shell instance using `|`

bitoiu commented on downloads/github-actions-cheat-sheet.md:
```diff
@@ -0,0 +1,193 @@
+layout: cheat-sheet
+redirect_to: false
+title: GitHub Actions Cheat Sheet
+byline: GitHub Actions give you the flexibility to build automated software development lifecycle workflows. You can write individual tasks, called actions, and combine them to create  custom workflows in your repository. GitHub Actions are automated processes allowing you to build, test, package, release, or deploy any code project on GitHub, but you can also use them to automate any step of you ...
```
```suggestion
byline: GitHub Actions give you the flexibility to build automated software development lifecycle workflows. You can write individual tasks, called actions, and combine them to create custom workflows in your repository. GitHub Actions are automated processes allowing you to build, test, package, release, or deploy any code project on GitHub, but you can also use them to automate any step of your workflow&#58; merging pull requests, assigning labels, triaging issues to name a few.
```

Removed extra whitespace, but what about the `workflow&#58`, I think the `;` is already there. Should we take it out?

helaili replied:
":" is a Jekyll delelimiter so I had to use the HTML code

bitoiu commented on downloads/github-actions-cheat-sheet.md:
```diff
@@ -0,0 +1,193 @@
+
+{% capture colOne %}
+## Workflow Syntax
+Workflow files use YAML syntax, and must have either a .yml or .yaml file extension. You must store workflow files in the `.github/workflows/` directory of your repository.
```
```suggestion
Workflow files use YAML syntax, and must have either a .yml or .yaml file extension. You must store workflow files in the `.github/workflows/` directory of your repository. Each different YAML file corresponds to a different workflow.
```

helaili commented:
Thanks for the reviews and comment. I’ve been working on an alternate way of generating a PDF, an automated way rather than copy/paste into an Adobe Illustrator file. I’ve made good progress so we should be able to focus again on the content soon.

helaili commented:
Closing in favor of  github/actions-cheat-sheet#5
